### PR TITLE
385 cli conversion to api part 2

### PR DESCRIPTION
**Pull request by Weebtrain**
## 📝 Description

Converted 2 more sections of CLI to API calls. Particularly in project insights. Difficulty was met in the abstraction of the code and a significant amount of tracing was needed to find what variables were.

More API conversion and tests are needed.

**Does not close:** #385

---

## 🔧 Type of Change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [ ] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📚 Documentation added/updated
- [x] ✅ Test added/updated
- [ ] ♻️ Refactoring
- [ ] ⚡ Performance improvement

---

## 🧪 Testing

Run all docker tests to confirm functionality.

Under test_project_insights_API.py
- [x] test_return_skill_insights_chronological()
- [x] test_return_project_insights_chronological()

---

## ✓ Checklist

- [ ] 🤖 GenAI was used in generating the code and I have performed a self-review of my own code
- [x] 💬 I have commented my code where needed
- [x] 📖 I have made corresponding changes to the documentation
- [x] ⚠️ My changes generate no new warnings
- [x] ✅ I have added tests that prove my fix is effective or that my feature works and tests are passing locally
- [ ] 🔗 Any dependent changes have been merged and published in downstream modules
- [ ] 📱 Any UI changes have been checked to work on desktop, tablet, and/or mobile

### CODE REVIEW

File: src/API/project_insights_API.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            ns.languages = p.get("languages", []) or []
+            ns.frameworks = p.get("frameworks", []) or []
+            ns.summary = p.get("summary", "")
+            normalized.append(ns)
+        projects = normalized
+    projects = filter_insights(
+        projects,
+        language=language,
+        skill=skill,
+        since=since_dt,
```
Comment (mahigangal) [inline src/API/project_insights_API.py]: GET /insights/projects can break when skill filter is used. 
In the dict-to-object conversion, the temp object does not include skills, but filter_insights() expects ins.skills.

Add ns.skills = p.get("skills", []) or [].

File: src/API/project_insights_API.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    projects_dicts = None
+    try:
+        preferred = apply_preferences()
+        projects = [
+            p
+            for p in preferred.get("projects", [])
+            if parse_date(p.get("analyzed_at")) is not None
+        ]
+    except FileNotFoundError:
+        projects = list_project_insights(storage_path=storage_path)
```
Comment (mahigangal) [inline src/API/project_insights_API.py]: This endpoint is supposed to be chronological, but when apply_preferences() is used, it can return user-preferred order instead of chronological order.

File: src/API/project_insights_API.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        skill=skill,
+        since=since_dt,
+    )
+    #cannot return namespaces or ProjectInsight objects through api calls so we need to convert to dicts
+    filtered_projects = []
+    if projects_dicts:
+        for p in projects:
+            for d in projects_dicts:
+                if d.get("project_name", "unknown") == p.project_name:
+                    filtered_projects.append(d)
```
Comment (ssmith86) [inline src/API/project_insights_API.py]: The nested loop at lines 71 to 74 is O(n²), which is not ideal. Since the namespaces are built from projects_dicts in the same order, you could just build a lookup dict upfront: dicts_by_name = {d.get("project_name"): d for d in projects_dicts} then filtered_projects = [dicts_by_name[p.project_name] for p in projects if p.project_name in dicts_by_name].

File: test/test_project_insights_API.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
+from fastapi.testclient import TestClient
```
[the remark this answers is not in the document]
Comment (ssmith86) [reply test/test_project_insights_API.py]: I second this: the existing tests only cover the happy path (count + ordering). There's no coverage for the language, skill, or since_str filter parameters. Consider adding tests for those, and populating the fixture with at least one entry that has non-empty languages/skills so the filters can be meaningfully exercised.

**Review by ssmith86 — CHANGES_REQUESTED** [review]
There's a couple of fixes that should be made before merging, but otherwise it looks good. 
Existing tests both pass, but coverage is a bit low. Please see comments for some ideas on what to add.
Image

File: src/API/project_insights_API.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Puneet-Maan) [reply src/API/project_insights_API.py]: @ssmith86 I checked the code and project_name isn’t guaranteed unique before this step, and the dedup logic is only for duplicate files, not duplicate insight records. So matching by name alone here can definitely return the wrong or extra entry if two insights share the same project name.

File: src/API/project_insights_API.py (hunk 2 above)
Comment (Weebtrain) [reply src/API/project_insights_API.py]: Removed it. Didn't know what it did but I copied it from Puneet's chronoloigcal listing

**Review by mahigangal — APPROVED** [review]

**Review by ssmith86 — APPROVED** [review]
All requested changes have been made. Everything looks good. As discussed, tests will be updated next week. Current tests still all pass

Image
